rosidl_typesupport_freertps_cpp/rosidl_typesupport_freertps_cpp/__init__.py

```diff
@@ -177,3 +91,624 @@ def generate_typesupport_freertps_cpp(args):
+{0}_p += sizeof(uint32_t);
+{0}for (uint32_t _{1}_idx = 0; _{1}_idx < _{1}_len; _{1}_idx++)
+{0}{{""".format(" " * indent, field_name)
+        output += enforce_alignment(4, indent=indent + 2)
+        output += """{0}  uint32_t _{1}_len = (uint32_t) _s->{1}[_{1}_idx].length() + 1;
+{0}  *((uint32_t *)_p) = _{1}_len;
+{0}  _p += sizeof(uint32_t);
+{0}  for (uint32_t i = 0; i < _{1}_len - 1; ++i)
+{0}  {{
+{0}    *((char*) _p) = _s->{1}[_{1}_idx][i];
+{0}    _p += sizeof(char);
+{0}  }}
+{0}  *((char*) _p) = '\\0';
+{0}  ++_p;
+{0}}}
+""".format(" " * indent, field_name)
+        return output
+
+    def deserialize_variable_array(self, field_name, indent=2):
+        output = enforce_read_alignment(4, indent)  # align for the sequence count
+        output += """
+{0}size_t _{1}_len = *((uint32_t *)*_p);
+{0}*_p += sizeof(uint32_t);
+{0}*_len -= sizeof(uint32_t);
+{0}if (_s->{1}.size() != _{1}_len)
+{0}{{
+{0}  _s->{1}.resize(_{1}_len);
+{0}}}
+{0}for (uint32_t _{1}_idx = 0; _{1}_idx < _{1}_len; _{1}_idx++)
+{0}{{""".format(" " * indent, field_name)
+        output += enforce_read_alignment(4, indent=indent + 2)
+        output += """{0}    uint32_t _{1}_len = *((uint32_t *)*_p);
+{0}  if (*_len < sizeof(uint32_t))
+{0}{{
+{0}    return false;
+{0}}}
+{0}  (*_p) += sizeof(uint32_t);
+{0}  (*_len) -= sizeof(uint32_t);
+{0}  _s->{1}[_{1}_idx].assign(_{1}_len-1, '\\0');
+{0}  for (uint32_t i = 0; i < _{1}_len - 1; ++i)
+{0}  {{
+{0}    if (*_len < sizeof(char))
+{0}    {{
+{0}      return false;
+{0}    }}
+{0}    _s->{1}[_{1}_idx][i] = *((char *)*_p);
+{0}    *_p += sizeof(char);
+{0}    *_len -= sizeof(char);
+{0}  }}
+{0}  ++(*_p);
+{0}  --(*_len);
+{0}}}
+""".format(" " * indent, field_name)
+        return output
+
+    def get_buffer_size(self, field_name, indent=2):
+        output = enforce_alignment_buffer_size(4, indent)
+        output += """
+{0}_len += sizeof(uint32_t);
+{0}_len += _s->{1}.size() + 1;
+""".format(" " * indent, field_name)
+        return output
+
+    def get_buffer_size_fixed_array(self, field_name, array_size, indent=2):
+        output = """
+{0}for (uint32_t _{1}_idx = 0; _{1}_idx < {2}; _{1}_idx++)
+{0}{{""".format(" " * indent, field_name, array_size)
+        output += enforce_alignment_buffer_size(4, indent=indent + 2)
+        output += """{0}  _len += sizeof(uint32_t);
+{0}  _len += _s->{1}[_{1}_idx].size() + 1;
+{0}}}
+""".format(" " * indent, field_name)
+        return output
+
+    def get_buffer_size_variable_array(self, field_name, indent=2):
+        output = enforce_alignment_buffer_size(4, indent)
+        output += """
+{0}_len += sizeof(uint32_t);
+{0}for (uint32_t _{1}_idx = 0; _{1}_idx < _s->{1}.size(); _{1}_idx++)
+{0}{{""".format(" " * indent, field_name)
+        output += enforce_alignment_buffer_size(4, indent=4)
+        output += """{0}  _len += sizeof(uint32_t);
+{0}  _len += _s->{1}[_{1}_idx].length() + 1;
+{0}}}
+""".format(" " * indent, field_name)
+        return output
```

Comment:
This embeds a lot of C++ code in the Python generator. This should be moved to the templates.